### DISCUSSION TITLE
AsyncCommand can execute questions

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [xamarin/XamarinCommunityToolkit#709] [Q&A] [answered]
1) Why does `canExecute` parameter for `AsyncCommand<T>` accept `Func<object, bool>` instead of `Func<T, bool>`? It's `Func<T, bool>` for regular `Command<T>`.
2) Why does `canExecute` parameter for `AsyncCommand` (without `T`)  accept `Func<object, bool>` instead of `Func<bool>`? It's `Func<bool>` for regular `Command`.
3) Why `RaiseCanExecuteChanged()` isn't named `ChangeCanExecute()` to be consistent with similar method from `Command`?
4) It seems like  `AsyncCommand<T>` should be `AsyncCommand<in T>` since IAsyncCommand is declared as `IAsyncCommand<in T>` and we don't return `T` from anywhere in `AsyncCommand<T>`

**Comment by jfversluis** [comment]
@brminnick you are probably the best person to answer this if I'm not mistaken?

**Comment by TheCodeTraveler** [comment]
No problem! 

Since some answers are quite verbose, I'll break them out into separate responses to help with readability 👍

**Comment by TheCodeTraveler** [comment, marked as answer]
## Question 1 

> Why does `canExecute` parameter for `AsyncCommand<T>` accept `Func<object, bool>` instead of `Func<T, bool>`? It's `Func<T, bool>` for regular `Command<T>`.

### Answer
 In `AsyncCommand<T>`, `T` is the type for the parameter for `AsyncCommand.Execute(T parameter)`. 

We [purposely did not use `T` for `canExecute<object, bool>`](https://github.com/xamarin/XamarinCommunityToolkit/pull/279#issuecomment-692882191) to allow for scenarios where `CanExecute` and and `Execute` use different parameters. For example, here is a case where we are using an `int` for the Execute parameter and a `bool` for the CanExecute parameter:

```csharp
public ICommand PullToRefreshCommand { get; } = new AsyncCommand((int) count => FetchNewData(count), (bool) isBusy => !isBusy)
```

This implementation best honors the paradigm of [`ICommand`](https://docs.microsoft.com/dotnet/api/system.windows.input.icommand?view=netcore-3.1&WT.mc_id=xamarincommunitytoolkit-github-bramin) which doesn't require `Execute` to use the same parameter type as `CanExecute`.

I strongly disagree with Xamarin.Forms' original implementation of `Command<T>` that assigns `T` to both `Execute` and `CanExecute`, because the [`ICommand Specification`](https://docs.microsoft.com/dotnet/api/system.windows.input.icommand?view=netcore-3.1&WT.mc_id=xamarincommunitytoolkit-github-bramin) does not require `Execute` and `CanExecute` to use the same parameter type.

### Recommendation

For a type-safe `canExecute`, I recommend using `AsyncCommand<TExecute, TCanExecute>` like so:

```csharp
public ICommand PullToRefreshCommand { get; } = new AsyncCommand<int,bool>(count => FetchNewData(count), isBusy => !isBusy)
```

**Reply by maxkoshevoi** [reply]
Oh git it, makes sense. But wouldn't it be better if `AsyncCommand<T>` was equivalent to `AsyncCommand<T, T>` and if I needed to have `canExecute` with different type, I would create command with two type arguments? This way we always have type-safe `canExecute` and an option for it to be different type if we need to.

**Reply by TheCodeTraveler** [reply]
It's a fair argument, but it'd be a breaking change to implement it, and the type-safe functionality already exists in `AsyncCommand<TExecute, TCanExecute>`.

This would be a tough one to advocate for since we'd be introducing a breaking change only to provide a benefit that already exists.

**Reply by maxkoshevoi** [reply]
Opt-in type-safe is not ideal.. This is new package (according to GitHub only 38 projects use it so far), so there is still time to introduce changes like this. Or this change can be merged on next major release.

If I would need type-safe `canExecute` (which I assume everyone coding with staticky typed language wants), I'll need to use `AsyncCommand<TExecute, TCanExecute>` anyway.

if I don't need can execute, I'll use `AsyncCommand<TExecute>` in both cases.

This change will be breaking only for those who are using not type-safe `canExecute` (maybe because they knew they could), so it would just point they to more cleaner code.

**Comment by TheCodeTraveler** [comment]
## Question 2

> Why does `canExecute` parameter for `AsyncCommand` (without `T`)  accept `Func<object, bool>` instead of `Func<bool>`? It's `Func<bool>` for regular `Command`.

### Answer

Similar to the [Answer for Question #1](https://github.com/xamarin/XamarinCommunityToolkit/discussions/709#discussioncomment-269961), `Execute` and `CanExecute` are separate and are not required by the [ICommand Specification](https://docs.microsoft.com/dotnet/api/system.windows.input.icommand?view=netcore-3.1&WT.mc_id=xamarincommunitytoolkit-github-bramin) to use the same parameter type. 

I.e. Foregoing an input parameter for `Execute` does not predicate that you must forego an input parameter for `CanExecute`.

Here is an example where `Execute` does not use an input parameter and `CanExecute` _does_ use an input parameter:

```csharp
public ICommand PullToRefreshCommand { get; } = new AsyncCommand(() => FetchNewData(count), (bool) isBusy => !isBusy)
```

Again, I strongly disagree with Xamarin.Forms' implementation of `Command` as it breaks the `ICommand` paradigm.

**Comment by TheCodeTraveler** [comment]
## Question 3

> Why `RaiseCanExecuteChanged()` isn't named `ChangeCanExecute()` to be consistent with similar method from `Command`?

### Answer

Good point! We should include `ChangeCanExecute()` to provide for an easier transition from `Command<T>` to `AsyncCommand<T>`.

The reason I chose the name `RaiseCanExecuteChanged`  is mostly a matter of semantics: [`RaiseCanExecuteChanged` invokes `ICommand. CanExecuteChanged`](https://github.com/xamarin/XamarinCommunityToolkit/blob/21bec3e669a78ea4c55f25985001cd73f8498db9/src/CommunityToolkit/Xamarin.CommunityToolkit/ObjectModel/BaseCommand.shared.cs#L82-L86); it allows you to Raise (or invoke) the `event CanExecute`.

**tl;dr** `RaiseCanExecuteChanged` is a better description than `ChangeCanExecute` of what is happening under the hood when the method is called.

### Resolution

I'll open a PR to add `ChangeCanExecute()` 🙌

**Reply by TheCodeTraveler** [reply]
### Update

Pull Request Opened: https://github.com/xamarin/XamarinCommunityToolkit/pull/736

**Reply by maxkoshevoi** [reply]
Yes, it sounds better, but backwards compatibility demands sacrifices 😄
In your PR you marked `ChangeCanExecute()` as `Obsolete`. Point of my question was that two Command classes should have the same interface so that if some method (that is used to create a command) becomes `async` (or sync) all programmer needs to do is to change `new Command`  to `new AsyncCommand`. And right now biggest differences that I noticed are `canExecute` type (discussed above, so not touching here) and different name of `ChangeCanExecute`.

**Reply by TheCodeTraveler** [reply]
Correct, and this PR provides backwards compatibility. 

I marked it as `[Obsolete]` to reduce the maintenance required for two methods that do the same thing, and to reduce the overhead of documenting two methods that do the same thing.

**Comment by TheCodeTraveler** [comment]
## Question 4

> It seems like  `AsyncCommand<T>` should be `AsyncCommand<in T>` since IAsyncCommand is declared as `IAsyncCommand<in T>` and we don't return `T` from anywhere in `AsyncCommand<T>`

## Answer 

Because `<in T>` is only allowed for Interfaces and Delegates. 

If we try to use `<in T>` for a class, we get the following compiler error:

<img width="1027" alt="image" src="https://user-images.githubusercontent.com/13558917/104056386-42f8fe00-51a5-11eb-9357-5c27cd7c6ad5.png">

The good news is that we are using `<in T>` for both [IAsyncCommand](https://github.com/xamarin/XamarinCommunityToolkit/blob/21bec3e669a78ea4c55f25985001cd73f8498db9/src/CommunityToolkit/Xamarin.CommunityToolkit/ObjectModel/IAsyncCommand.shared.cs#L7) and [IAsyncValueCommand](https://github.com/xamarin/XamarinCommunityToolkit/blob/21bec3e669a78ea4c55f25985001cd73f8498db9/src/CommunityToolkit/Xamarin.CommunityToolkit/ObjectModel/IAsyncCommand.shared.cs#L7)!

**Reply by maxkoshevoi** [reply]
Oh, I missed that part of C#, sorry. Thanks for pointing it out!
